### ISSUE TITLE
Meaning of `query_only` in setup.py?

### ISSUE THREAD

**Issue by 88d52bdba0366127fffca9dfa93895** [mrjbq7/ta-lib#322]
Hi @mrjbq7,

Building talib on my remote server in Dockerfile it give me an error as line `import numpy`, I fixed it and for me, it's not a problem. But I just curious about your purpose, I believe that we could remove `query_only` (and two related lines) without any problem.

ref: https://github.com/mrjbq7/ta-lib/blob/master/setup.py#L61

**Comment by mrjbq7** [comment]
It had to do with setuptools listing dependencies (like numpy) which would be installed prior to running the installation process (which uses numpy for the include directories).

What was the error you got?

**Comment by 88d52bdba0366127fffca9dfa93895** [comment]
Your [`import numpy`](https://github.com/mrjbq7/ta-lib/blob/master/setup.py#L62) makes us have to install `numpy` before we can install this library, for example:

https://github.com/deepnox-io/docker-python-ta-lib/blob/master/Dockerfile#L28
https://github.com/quantopian/zipline/blob/master/Dockerfile#L57

Moreover, the installation should be simple such as `pip install TA-Lib` and all dependencies are handled by itself, then user **dont have to** know they need to install `numpy` first. Inspired by `pandas`, which also depends on `numpy` and `cython`, [here](https://github.com/pandas-dev/pandas/blob/master/setup.py#L390) we could bypass over this restriction. Another implementation from `zipline` [here](https://github.com/quantopian/zipline/blob/master/setup.py#L36) is very impressive. I give it a try and will give you the output below.

**Comment by 88d52bdba0366127fffca9dfa93895** [comment]
hi @mrjbq7, sorry for late reply. Here is my try which contains multiple query (e.g. `egg_info`, `--help`) and install TA-Lib without numpy, finally is your test cases (using `nosetests`). The result is promising and we can use this way to setup the package. I post a long comment here to let you review it, please let me know if you need to check more commands. I also pushed the modification into my fork, you can take a look at [here](https://github.com/88d52bdba0366127fffca9dfa93895/ta-lib/commit/0bd5d937955542a3e08905625d06c9871f32bb3d). 

```diff
diff --git a/setup.py b/setup.py
index 4ed0c12..cf359c2 100644
--- a/setup.py
+++ b/setup.py
@@ -6,12 +6,15 @@ import warnings
 
 from distutils.dist import Distribution
 
-display_option_names = Distribution.display_option_names + ['help', 'help-commands']
-query_only = any('--' + opt in sys.argv for opt in display_option_names) or len(sys.argv) < 2 or sys.argv[1] == 'egg_info'
+# display_option_names = Distribution.display_option_names + ['help', 'help-commands']
+# query_only = any('--' + opt in sys.argv for opt in display_option_names) or len(sys.argv) < 2 or sys.argv[1] == 'egg_info'
 
 try:
     from setuptools import setup, Extension
-    requires = {"install_requires": ["numpy"]}
+    requires = {
+        "install_requires": ["numpy", 'cython'],
+        "setup_requires": ['numpy', 'cython']
+    }
 except:
     from distutils.core import setup
     from distutils.extension import Extension
@@ -57,16 +60,16 @@ if sys.platform == "win32":
 if not platform_supported:
     raise NotImplementedError(sys.platform)
 
-# Do not require numpy or cython for just querying the package
-if not query_only:
-    import numpy
-    include_dirs.insert(0, numpy.get_include())
-
-try:
-    from Cython.Distutils import build_ext
-    has_cython = True
-except ImportError:
-    has_cython = False
+# # Do not require numpy or cython for just querying the package
+# if not query_only:
+#     import numpy
+#     include_dirs.insert(0, numpy.get_include())
+#
+# try:
+#     from Cython.Distutils import build_ext
+#     has_cython = True
+# except ImportError:
+#     has_cython = False
 +# # Do not require numpy or cython for just querying the package
+# if not query_only:
+#     import numpy
+#     include_dirs.insert(0, numpy.get_include())
+#
+# try:
+#     from Cython.Distutils import build_ext
+#     has_cython = True
+# except ImportError:
+#     has_cython = False
 
 for lib_talib_dir in lib_talib_dirs:
     try:
@@ -78,30 +81,77 @@ for lib_talib_dir in lib_talib_dirs:
 else:
     warnings.warn('Cannot find ta-lib library, installation may fail.')
 
-cmdclass = {}
-if has_cython:
-    cmdclass['build_ext'] = build_ext
+# cmdclass = {}
+# if has_cython:
+#     cmdclass['build_ext'] = build_ext
+
+
+class LazyBuildExtCommandClass(dict):
+    """
+    Lazy command class that defers operations requiring Cython and numpy until
+    they've actually been downloaded and installed by setup_requires.
+    """
+
+    def __contains__(self, key):
+        return (key == 'build_ext' or
+                super(LazyBuildExtCommandClass, self).__contains__(key))
+
+    def __setitem__(self, key, value):
+        if key == 'build_ext':
+            raise AssertionError("build_ext overridden!")
+        super(LazyBuildExtCommandClass, self).__setitem__(key, value)
+
+    def __getitem__(self, key):
+        global include_dirs
+        if key != 'build_ext':
+            return super(LazyBuildExtCommandClass, self).__getitem__(key)
+
+        from Cython.Distutils import build_ext as cython_build_ext
+        import numpy
+
+        # Cython_build_ext isn't a new-style class in Py2.
+        class build_ext(cython_build_ext, object):
+            """
+            Custom build_ext command that lazily adds numpy's include_dir to
+            extensions.
+            """
+
+            def build_extensions(self):
+                """
+                Lazily append numpy's include directory to Extension includes.
+                This is done here rather than at module scope because setup.py
+                may be run before numpy has been installed, in which case
+                importing numpy and calling `numpy.get_include()` will fail.
+                """
+                numpy_incl = numpy.get_include()
+                for ext in self.extensions:
+                    ext.include_dirs.append(numpy_incl)
+
+                super(build_ext, self).build_extensions()
+
+        return build_ext
+
+
+cmdclass = LazyBuildExtCommandClass()
 
 ext_modules = [
     Extension(
         'talib._ta_lib',
-        ['talib/_ta_lib.pyx' if has_cython else 'talib/_ta_lib.c'],
+        ['talib/_ta_lib.pyx'],  # if has_cython else 'talib/_ta_lib.c'],
         include_dirs=include_dirs,
         library_dirs=lib_talib_dirs,

```

And here is a Dockerfile to reproduce my try

```Dockerfile
# Dockerfile for TA-Lib. To build:
#
#    docker build --rm -t talib -f Dockerfile .
#
# To run:
#
#    docker run --rm -it talib bash
#
FROM python:3.6

RUN apt-get update && apt-get install -y \
    gfortran \
    libfreetype6-dev \
    libhdf5-dev \
    liblapack-dev \
    libopenblas-dev \
    libpng-dev \
  && rm -rf /var/lib/apt/lists/* \
  && curl -L http://prdownloads.sourceforge.net/ta-lib/ta-lib-0.4.0-src.tar.gz \
  | tar xvz \
  && cd /ta-lib \
  && ./configure --prefix=/usr \
  && make \
  && make install \
  && pip install --upgrade pip

WORKDIR /TA-Lib
COPY . .

RUN python setup.py egg_info \
  && python setup.py --help \
  && pip install -e . \
  && pip -V \
  && pip freeze \
  && pip show numpy cython TA_Lib \
  && python -c 'import numpy; import talib; close = numpy.random.random(100); output = talib.SMA(close); print(output)' \
  && pip install -r requirements_dev.txt \
  && pip install -r requirements_test.txt \
  && nosetests -w .

```

And below is the output of step 5 in Dockerfile

```
Step 5/5 : RUN python setup.py egg_info   && python setup.py --help   && pip install -e .   && pip -V   && pip freeze   && pip show numpy cython TA_Lib   && python -c 'import numpy; import talib; close = numpy.random.random(100); output = talib.SMA(close); print(output)'   && pip install -r requirements_dev.txt   && pip install -r requirements_test.txt   && nosetests -w .
 ---> Running in 31bc5a7cb239
running egg_info
creating TA_Lib.egg-info
writing TA_Lib.egg-info/PKG-INFO
writing dependency_links to TA_Lib.egg-info/dependency_links.txt
writing requirements to TA_Lib.egg-info/requires.txt
writing top-level names to TA_Lib.egg-info/top_level.txt
writing manifest file 'TA_Lib.egg-info/SOURCES.txt'
reading manifest file 'TA_Lib.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'TA_Lib.egg-info/SOURCES.txt'
Common commands: (see '--help-commands' for more)

  setup.py build      will build the package underneath 'build/'
  setup.py install    will install the package

Global options:
  --verbose (-v)      run verbosely (default)
  --quiet (-q)        run quietly (turns verbosity off)
  --dry-run (-n)      don't actually do anything
  --help (-h)         show detailed help message
  --no-user-cfg       ignore pydistutils.cfg in your home directory
  --command-packages  list of packages that provide distutils commands

Information display options (just display information, ignore any commands)
  --help-commands     list all available commands
  --name              print package name
  --version (-V)      print package version
  --fullname          print <package name>-<version>
  --author            print the author's name
  --author-email      print the author's email address
  --maintainer        print the maintainer's name
  --maintainer-email  print the maintainer's email address
  --contact           print the maintainer's name if known, else the author's
  --contact-email     print the maintainer's email address if known, else the
                      author's
  --url               print the URL for this package
  --license           print the license of the package
  --licence           alias for --license
  --description       print the package description
  --long-description  print the long package description
  --platforms         print the list of platforms
  --classifiers       print the list of classifiers
  --keywords          print the list of keywords
  --provides          print the list of packages/modules provided
  --requires          print the list of packages/modules required
  --obsoletes         print the list of packages/modules made obsolete

usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: setup.py --help [cmd1 cmd2 ...]
   or: setup.py --help-commands
   or: setup.py cmd --help

Obtaining file:///TA-Lib
Collecting numpy
  Using cached numpy-1.18.5-cp36-cp36m-manylinux1_x86_64.whl (20.1 MB)
Collecting cython
  Using cached Cython-0.29.20-cp36-cp36m-manylinux1_x86_64.whl (2.0 MB)
Installing collected packages: numpy, cython, TA-Lib
  Running setup.py develop for TA-Lib
Successfully installed TA-Lib cython-0.29.20 numpy-1.18.5
pip 20.1.1 from /usr/local/lib/python3.6/site-packages/pip (python 3.6)
Cython==0.29.20
numpy==1.18.5
-e git+https://github.com/88d52bdba0366127fffca9dfa93895/ta-lib@633c4d349e64b136b98dea15460e34de4f1cf4a4#egg=TA_Lib
Name: numpy
Version: 1.18.5
Summary: NumPy is the fundamental package for array computing with Python.
Home-page: https://www.numpy.org
Author: Travis E. Oliphant et al.
Author-email: None
License: BSD
Location: /usr/local/lib/python3.6/site-packages
Requires: 
Required-by: TA-Lib
---
Name: Cython
Version: 0.29.20
Summary: The Cython compiler for writing C extensions for the Python language.
Home-page: http://cython.org/
Author: Robert Bradshaw, Stefan Behnel, Dag Seljebotn, Greg Ewing, et al.
Author-email: cython-devel@python.org
License: Apache
Location: /usr/local/lib/python3.6/site-packages
Requires: 
Required-by: TA-Lib
---
Name: TA-Lib
Version: 0.4.18
Summary: Python wrapper for TA-Lib
Home-page: http://github.com/mrjbq7/ta-lib
Author: John Benediktsson
Author-email: mrjbq7@gmail.com
License: UNKNOWN
Location: /TA-Lib
Requires: numpy, cython
Required-by: 
[       nan        nan        nan        nan        nan        nan
        nan        nan        nan        nan        nan        nan
        nan        nan        nan        nan        nan        nan
        nan        nan        nan        nan        nan        nan
        nan        nan        nan        nan        nan 0.47184421
 0.47579143 0.48807117 0.49353794 0.50692695 0.5130363  0.50641668
 0.4841212  0.48034287 0.47944732 0.50595572 0.49483965 0.48342809
 0.50561175 0.49735046 0.50727258 0.49122906 0.48507554 0.47828914
 0.48626616 0.50609962 0.50452747 0.50075265 0.47973105 0.48266352
 0.47313579 0.46250675 0.4604351  0.48094534 0.46771316 0.48510767
 0.4682141  0.47272184 0.4738814  0.48317955 0.4980949  0.50040955
 0.49544414 0.50592702 0.5323088  0.51357514 0.51412836 0.53180067
 0.51196771 0.51438223 0.50108429 0.50472314 0.51721719 0.50967181
 0.49056284 0.4796205  0.48088973 0.48350688 0.51051277 0.52006705
 0.5215563  0.52016046 0.54058168 0.51915099 0.53279052 0.50833878
 0.50315686 0.50073458 0.51955109 0.50666772 0.50101805 0.49693754
 0.49127585 0.4972008  0.48004255 0.50071604]
Collecting numpy==1.16.2
  Downloading numpy-1.16.2-cp36-cp36m-manylinux1_x86_64.whl (17.3 MB)
Collecting Cython==0.29.7
  Downloading Cython-0.29.7-cp36-cp36m-manylinux1_x86_64.whl (2.1 MB)
Collecting beautifulsoup4==4.7.1
  Downloading beautifulsoup4-4.7.1-py3-none-any.whl (94 kB)
Collecting mistune==0.8.4
  Downloading mistune-0.8.4-py2.py3-none-any.whl (16 kB)
Collecting Pygments==2.3.1
  Downloading Pygments-2.3.1-py2.py3-none-any.whl (849 kB)
Collecting soupsieve>=1.2
  Downloading soupsieve-2.0.1-py3-none-any.whl (32 kB)
Installing collected packages: numpy, Cython, soupsieve, beautifulsoup4, mistune, Pygments
  Attempting uninstall: numpy
    Found existing installation: numpy 1.18.5
    Uninstalling numpy-1.18.5:
      Successfully uninstalled numpy-1.18.5
  Attempting uninstall: Cython
    Found existing installation: Cython 0.29.20
    Uninstalling Cython-0.29.20:
      Successfully uninstalled Cython-0.29.20
Successfully installed Cython-0.29.7 Pygments-2.3.1 beautifulsoup4-4.7.1 mistune-0.8.4 numpy-1.16.2 soupsieve-2.0.1
Requirement already satisfied: numpy==1.16.2 in /usr/local/lib/python3.6/site-packages (from -r requirements.txt (line 1)) (1.16.2)
Requirement already satisfied: Cython==0.29.7 in /usr/local/lib/python3.6/site-packages (from -r requirements.txt (line 2)) (0.29.7)
Collecting pandas==0.24.2
  Downloading pandas-0.24.2-cp36-cp36m-manylinux1_x86_64.whl (10.1 MB)
Collecting nose==1.3.7
  Downloading nose-1.3.7-py3-none-any.whl (154 kB)
Collecting python-dateutil>=2.5.0
  Downloading python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)
Collecting pytz>=2011k
  Downloading pytz-2020.1-py2.py3-none-any.whl (510 kB)
Collecting six>=1.5
  Downloading six-1.15.0-py2.py3-none-any.whl (10 kB)
Installing collected packages: six, python-dateutil, pytz, pandas, nose
Successfully installed nose-1.3.7 pandas-0.24.2 python-dateutil-2.8.1 pytz-2020.1 six-1.15.0
.....................................
----------------------------------------------------------------------
Ran 37 tests in 0.502s

OK

```

**Comment by mrjbq7** [comment]
That looks great!

I had previously made ``cython`` an optional requirement, by bundling the generated ``C`` files we don't need cython to regenerate them before building/installing.

I assume with the new code, that could be done where we import ``cython_build_ext`` and make that optional?

**Comment by 88d52bdba0366127fffca9dfa93895** [comment]
So, your `talib/_ta_lib.pyx` file is generated from `talib/_ta_lib.c` and the .c file is the main source, and now you dont want to do it again? Just install from .c file? If yes, you can replace the `.pyx` to `.c` in this [line](https://github.com/88d52bdba0366127fffca9dfa93895/ta-lib/commit/0bd5d937955542a3e08905625d06c9871f32bb3d#diff-2eeaed663bd0d25b7e608891384b7298R140) and try to run Dockerfile. (I tried and the output is the same)

**Comment by briancappello** [comment]
Almost, it's the other way around - cython generates the C files from the pyx/pxi files :) the flow is outlined a bit in the `DEVELOPMENT` file in the root of this repo

**Comment by 88d52bdba0366127fffca9dfa93895** [comment]
Ah got it, thanks @briancappello. So with the `cython_build_ext` we dont need the C files, just keep only pyx/pxi files. How do you think about this @mrjbq7?

**Comment by mrjbq7** [comment]
Several people have wanted to be able to install without making Cython a dependency. That’s why all that has_cython stuff was added.

**Comment by 88d52bdba0366127fffca9dfa93895** [comment]
Ok @mrjbq7, I modify and create a PR, a new version will use Cython if it's exist, please take a look at my PR. You can add `pip install cython` at this [line](https://github.com/mrjbq7/ta-lib/pull/324/files#diff-3254677a7917c6c01f55212f86c57fbfR30).

P/s: Dockerfile contains multiple lines to check/debug the installation, if you verify the PR I will remove redundant lines. I saw another PR contribute Dockerfile but IMO it's not good enough or at least need to be verified.
